tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/ral_legalize_to_llvm.cc

```diff
@@ -0,0 +1,556 @@
+  MLIRContext* ctx = builder.getContext();
+  Value globalPtr = builder.create<LLVM::AddressOfOp>(loc, globalOp);
+  Value cst0 = builder.create<LLVM::ConstantOp>(
+      loc, IntegerType::get(ctx, 64),
+      builder.getIntegerAttr(builder.getIndexType(), 0));
+  return builder.create<LLVM::GEPOp>(
+      loc, LLVM::LLVMPointerType::get(IntegerType::get(ctx, 8)), globalPtr,
+      ValueRange{cst0, cst0});
+}
+
+// Creates a global const string op named `name` using the value if not exists
+// and returns the Loaded value of this global op.
+Value loadOrCreateGlobalString(PatternRewriter& rewriter, Operation* op,
+                               StringRef name, StringRef value) {
+  ModuleOp module = op->getParentOfType<ModuleOp>();
+  GlobalOp globalOp = module.lookupSymbol<GlobalOp>(name);
+  if (!globalOp) {
+    OpBuilder::InsertionGuard guard(rewriter);
+    // Double-check under the lock
+    if (globalOp = module.lookupSymbol<GlobalOp>(name)) {
+      return loadGlobalString(rewriter, op->getLoc(), globalOp);
+    }
+    OpBuilder::InsertPoint ip = rewriter.saveInsertionPoint();
+    rewriter.setInsertionPointToStart(module.getBody());
+
+    auto type = LLVM::LLVMArrayType::get(IntegerType::get(op->getContext(), 8),
+                                         value.size());
+    globalOp = rewriter.create<LLVM::GlobalOp>(
+        op->getLoc(), type, /*isConstant=*/true, LLVM::Linkage::Internal, name,
+        rewriter.getStringAttr(value), /*alignment=*/0);
+
+    rewriter.restoreInsertionPoint(ip);
+  }
+
+  return loadGlobalString(rewriter, op->getLoc(), globalOp);
+}
+
+// Converts a ral.dispatch_op to its llvm format.
+struct DispatchOpToLLVMPattern : ConvertOpToLLVMPattern<DispatchOp> {
+  using ConvertOpToLLVMPattern<DispatchOp>::ConvertOpToLLVMPattern;
+
+  // Returns the ral dispatch function and inserts the declaration if not found.
+  LLVMFuncOp getOrInsertDispatchFunction(PatternRewriter& rewriter,
+                                         Operation* op) const;
+
+  // Packs the inputs and outputs into a type-erased pointer array.
+  // For example, `int func(int)` -> `void func(void* args[]) where args =
+  // {in_ptr, out_ptr}`
+  Value rewriteInsOutsOfDispatchOp(DispatchOp dispatch_op,
+                                   ArrayRef<Value> operands,
+                                   ConversionPatternRewriter& rewriter,
+                                   SmallVectorImpl<Value>& resultPtrs) const;
+
+  LogicalResult matchAndRewrite(
+      DispatchOp dispatch_op, ArrayRef<Value> operands,
+      ConversionPatternRewriter& rewriter) const override;
+};
+
+// Returns the llvm function definition of ral dispatch op and creates it first
+// if not exists.
+LLVMFuncOp DispatchOpToLLVMPattern::getOrInsertDispatchFunction(
+    PatternRewriter& rewriter, Operation* op) const {
+  ModuleOp module = op->getParentOfType<ModuleOp>();
+  LLVMFuncOp func = module.lookupSymbol<LLVMFuncOp>(kRalDispatchFunctionName);
+  if (!func) {
```

Comment:
In general, prefer early exit if possible: 
```suggestion
  if (func) return func;
```

Reply:
Done